### ISSUE TITLE
Sarif Logs cannot be split by runs (using SplittingStrategy.ByRun)

### ISSUE THREAD

**Issue by harleenkohli** [microsoft/sarif-sdk#2196]
This option is currently not implemented ref: https://github.com/microsoft/sarif-sdk/blob/main/src/Sarif/ExtensionMethods.cs 

Our team is looking at the option to use partitioningVisitor to split it ourselves as a work around.

**Comment by eddynaka** [comment]
@michaelcfanning , I was looking at this and I draft the following:
```csharp
case SplittingStrategy.PerRun:
{
    var dictionary = new Dictionary<string, string>();
    foreach (Run run in sarifLog.Runs)
    {
        if (run.Results != null)
        {
            foreach (Result result in run.Results)
            {
                dictionary[result.ResolvedRuleId(run)] = run.Tool.Driver.Name;
            }
        }
    }

    partitionFunction = (result) => dictionary[result.RuleId];
    break;
}
```

With this, we would create a dictionary containing Rule.Id => Tool Name. And, in the partition, that would be called.

Issues that I can see: if we have two runs at least and both have the same ruleid, the tool will get the wrong tool name

**Comment by michaelcfanning** [comment]
I don't think you need this. Instead, just use the run reference as your partition function. run.HashCode() should work.

**Comment by eddynaka** [comment]
The only issue with the partitionFunction is that it only accepts Result returning <T>. To do that, we would need to create a big refactor in the `PartitioningVisitor`.

**Comment by michaelcfanning** [comment]
The partition function is generic. You could actually use PartitionFunction<Run> to do your work. 1. Set the run on all results. 2. Have the predicate function return that run for each result. 

See `Run.SetRunOnResults()` helper.

**Comment by eddynaka** [comment]
I see! With that, we can simplify to this:

```csharp

                case SplittingStrategy.PerRun:
                {
                    foreach (Run run in sarifLog.Runs)
                    {
                        run.SetRunOnResults();
                    }

                    partitionFunction = (result) => result.Run.Tool.Driver.Name;
                    break;
                }
```

**Comment by michaelcfanning** [comment]
No, this won't work. This organizes all results across all runs by tool name. The identifier you need to return needs to be run-specific, not cross runs. Make sense? 

~~~
partitionFunction = (result) => result.Run.GetHashCode.ToString();
~~~

**Comment by eddynaka** [comment]
Got it. If we have two runs with the same name, in my case, it would gather everything together. And in yours, no.

Thanks!

I will create a few tests and open a PR 👍

**Comment by michaelcfanning** [comment]
Exactly! Now, there is a useful scenario for *merging* results from runs from the same tool into a single run. We need this for the PREfast C++ tool. It generates a log file per translation unit at compile/analysis time. The runs in all these logs are configured identically, so it makes sense to merge all these runs into a single run (in order to get a sense of the complete result set).

I mention this only to note that splitting/merging are two sides of the partitioning problem. We should clean up our SDK to get this in good shape. :)

@yongyan-gh
